pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -0,0 +1,496 @@
+		if errors.IsNotFound(err) {
+			// the Propagator can add virtual node for owners that
+			// don't exist yet, so we need to enqueue a virtual Delete event to
+			// remove the virtual node.
+			glog.V(6).Infof("item %v not found, generating a virtual delete event", item.identity)
+			event := event{
+				eventType: deleteEvent,
+				obj:       objectReferenceToUnstructured(item.identity),
+			}
+			gc.propagator.eventQueue.Add(event)
+			return nil
+		}
+		return err
+	}
+	if latest.GetUID() != item.identity.UID {
+		glog.V(6).Infof("UID doesn't match, item %v not found, ignore it", item.identity)
+		return nil
+	}
+	ownerReferences := latest.GetOwnerReferences()
+	if len(ownerReferences) == 0 {
+		glog.V(6).Infof("object %s's doesn't have an owner, continue on next item", item.identity)
+		return nil
+	}
+	// TODO: we need to remove dangling references if the object is not to be
+	// deleted.
+	for _, reference := range ownerReferences {
+		// TODO: we need to verify the reference resource is supported by the
+		// system. If it's not a valid resource, the garbage collector should i)
+		// ignore the reference when decide if the object should be deleted, and
+		// ii) should update the object to remove such references. This is to
+		// prevent objects having references to an old resource from being
+		// deleted during a cluster upgrade.
+		fqKind := unversioned.FromAPIVersionAndKind(reference.APIVersion, reference.Kind)
+		client, err := gc.clientPool.ClientForGroupVersion(fqKind.GroupVersion())
+		if err != nil {
+			return err
+		}
+		resource, err := gc.apiResource(reference.APIVersion, reference.Kind, len(item.identity.Namespace) != 0)
+		if err != nil {
+			return err
+		}
+		owner, err := client.Resource(resource, item.identity.Namespace).Get(reference.Name)
+		if err == nil {
+			if owner.GetUID() != reference.UID {
+				glog.V(6).Infof("object %s's owner %s/%s, %s is not found, UID mismatch", item.identity.UID, reference.APIVersion, reference.Kind, reference.Name)
+				continue
+			}
+			glog.V(6).Infof("object %s has at least an existing owner, will not garbage collect", item.identity.UID)
+			return nil
+		} else if errors.IsNotFound(err) {
+			glog.V(6).Infof("object %s's owner %s/%s, %s is not found", item.identity.UID, reference.APIVersion, reference.Kind, reference.Name)
+		} else {
+			return err
+		}
+	}
+	glog.V(2).Infof("none of object %s's owners exist any more, will garbage collect it", item.identity)
+	return gc.deleteObject(item.identity)
+}
+
+func (gc *GarbageCollector) Run(workers int, stopCh <-chan struct{}) {
+	for _, monitor := range gc.monitors {
+		go monitor.controller.Run(stopCh)
+	}
+
+	// worker
+	go wait.Until(gc.propagator.processEvent, 0, stopCh)
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(gc.worker, 0, stopCh)
+	}
+	<-stopCh
+	glog.Infof("Shutting down garbage collector")
+	gc.dirtyQueue.ShutDown()
+	gc.propagator.eventQueue.ShutDown()
+}
+
+func (gc *GarbageCollector) QueuesDrained() bool {
```

Comment:
Well, maybe you could make shut down block until the queues are completely
empty?

On Thu, May 12, 2016 at 1:19 PM, Chao Xu <notifications@github.com> wrote:

> In pkg/controller/garbagecollector/garbagecollector.go
> <https://github.com/kubernetes/kubernetes/pull/24509#discussion_r63089767>
> :
>
> > +		go monitor.controller.Run(stopCh)
> > +	}
> > +
> > +	// worker
> > +	go wait.Until(gc.propagator.processEvent, 0, stopCh)
> > +
> > +	for i := 0; i < workers; i++ {
> > +		go wait.Until(gc.worker, 0, stopCh)
> > +	}
> > +	<-stopCh
> > +	glog.Infof("Shutting down garbage collector")
> > +	gc.dirtyQueue.ShutDown()
> > +	gc.propagator.eventQueue.ShutDown()
> > +}
> > +
> > +func (gc *GarbageCollector) QueuesDrained() bool {
>
> @lavalamp <https://github.com/lavalamp> This function and the GraphHasUID
> function are used by the integration test. I don't want to expose them to
> users, do you have any suggestions?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24509/files/510f27d865886e11fa8543d89e7c0609576aaac3#r63089767>
>